letsencrypt-nginx/letsencrypt_nginx/nginxparser.py

```diff
@@ -1,56 +1,241 @@
-    space = White().suppress()
-    key = Word(alphanums + "_/")
-    # Matches anything that is not a special character AND any chars in single
-    # or double quotes
-    value = Regex(r"((\".*\")?(\'.*\')?[^\{\};,]?)+")
-    location = CharsNotIn("{};," + string.whitespace)
-    # modifier for location uri [ = | ~ | ~* | ^~ ]
-    modifier = Literal("=") | Literal("~*") | Literal("~") | Literal("^~")
-
-    # rules
-    comment = Literal('#') + restOfLine()
-    assignment = (key + Optional(space + value, default=None) + semicolon)
-    location_statement = Optional(space + modifier) + Optional(space + location)
-    if_statement = Literal("if") + space + Regex(r"\(.+\)") + space
-    block = Forward()
-
-    block << Group(
-        (Group(key + location_statement) ^ Group(if_statement)) +
-        left_bracket +
-        Group(ZeroOrMore(Group(comment | assignment) | block)) +
-        right_bracket)
-
-    script = OneOrMore(Group(comment | assignment) ^ block) + stringEnd
+    """A class that parses nginx configuration."""
 
     def __init__(self, source):
         self.source = source
 
-    def parse(self):
-        """Returns the parsed tree."""
-        return self.script.parseString(self.source)
-
     def as_list(self):
         """Returns the parsed tree as a list."""
-        return self.parse().asList()
+        return self.parse()
 
+    def parse(self):
+        """Returns the parsed tree."""
+        return RawNginxParser.Internal(self.source).parse_file()
+
+    class Internal(object):
+        """An internal class to keep track of parsing state."""
+        def __init__(self, source):
+            self.src = StringIO.StringIO(source)
+            self.peeked = None
+            self.eof = False
+
+        def _peek(self):
+            if self.eof:
+                return None
+
+            if self.peeked is None:
+                self.peeked = self.src.read(1)
+                if self.peeked == '':
+                    self.eof = True
+                    return None
+            return self.peeked
+
+        def _read(self):
+            if self.eof:
+                return None
+
+            if self.peeked:
+                val = self.peeked
+                self.peeked = None
+                return val
+            val = self.src.read(1)
+            if val == '':
+                self.eof = True
+                return None
+            return val
+
+        def parse_file(self):
+            """Tries to parse the file and returns a list of blocks parsed."""
+
+            result = self._parse_block()
+            if not self.eof:
+                raise ParseException("Invalid configuration file")
+
+            return result
+
+        def _parse_block(self):
+            result = []
+            while True:
+                res = self._parse_next()
+                if res is None:
+                    return result
+                result.append(res)
+
+        def _is_whitespace(self, val, includeNL):
```

Comment:
Let's be consistent with `_read_newline` and `_is_at_newline` here and call this `include_newline`.